### PR TITLE
Windows 7 support

### LINKED ISSUE #10085: Crystal programs doesn't work on Windows 7 due to GetCurrentThreadStackLimits
Right now the programs compiled with Crystal for Windows will only work on Windows 8 and higher.
The only cause (as far as I know) is the usage of function `GetCurrentThreadStackLimits`.
There is a way to get stack lower limit on win7(and lower), that is described here: https://stackoverflow.com/a/52404641
I understand that Windows 7 is outdated, but in this case just a simple patch will really increase availability of crystal programs.

### PR DESCRIPTION
**Pull request by konovod**
Closes #10085
This was discussed on [forum](https://forum.crystal-lang.org/t/will-there-be-cross-compile-support-for-windows-7/3168), but i haven't seen actual PR, so rolled my own. Still haven't checked on actual Windows 7, but at least produced file doesn't reference offending functions.

### CODE REVIEW

**Comment by straight-shoota** [comment]
There is no `win7` flag and I don't think it makes much sense to build different binaries for different windows versions. They should all use the same binary, but conduct runtime checks for which API is available.

**Comment by konovod** [comment]
I agree, but for now i'm not sure about how to check it. Is `GetProcAddress` at runtime for these two functions instead of static linking acceptable?

**Comment by oprypin** [comment]
FWIW if it helps someone and they really know they want Win7 and they don't mind passing `-Dwin7`, I don't see an issue with accepting this.
Someone really *should* try that this actually helps, though.

**Comment by konovod** [comment]
I can confirm that games built with -Dwin7 runs on Windows 7 just fine. Still can't check compiler building itself though.

**Comment by HertzDevil** [comment]
The problem with `flag?(:win7)` is that a new flag is required for each specific OS version. Instead, let's try to mimic the Windows SDK's definitions: (this is very similar to the recently added `LibC::ANDROID_API`)

```crystal
# c/sdkddkver.cr
lib LibC
  WIN32_WINNT_WIN7  = 0x0601
  WIN32_WINNT_WIN8  = 0x0602
  WIN32_WINNT_WIN10 = 0x0A00

  # add other version flags here, or use mechanisms other than flags
  {% if flag?(:win7) %}
    WIN32_WINNT = WIN32_WINNT_WIN7
  {% else %}
    WIN32_WINNT = WIN32_WINNT_WIN10
  {% end %}
end

# c/processthreadsapi.cr
lib LibC
  {% if LibC::WIN32_WINNT >= LibC::WIN32_WINNT_WIN8 %}
    fun GetCurrentThreadStackLimits(lowLimit : ULONG_PTR*, highLimit : ULONG_PTR*)
  {% end %}
end

{% if LibC.has_method?("GetCurrentThreadStackLimits") %}
  # ...
{% else %}
  # use the code in this PR
{% end %}
```

[`NTDDI_VERSION`](https://learn.microsoft.com/en-us/windows/win32/winprog/using-the-windows-headers) is even more fine-tuned and can detect things like Service Packs and the Creators Update.

**Comment by straight-shoota** [comment]
@HertzDevil Shelling out into the host system calls for trouble when cross-compiling. 
I suppose the current solution to determine the SDK level based on the `win7` flag is acceptable.
Maybe an alternative could be to allow configuration via an environment variable (`WIN32_WINNT`) and when that's unavailable fall back to shelling out (and default to the lowest one if that fails).
This would allow to specify the intended level explicitly for use cases such as cross-compiling.

**Comment by konovod** [comment]
Looks like `cmd.exe` isn't working in github CI, and `systeminfo` doesn't work too (systeminfo is also too slow, 1.5 seconds to every compilation imho isn't worth it). There are powershell-specific ways of detecting windows version, but then we lose compatibility with `cmd`.

**Review by oprypin — APPROVED** [review]

File: src/lib_c/x86_64-windows-msvc/c/sysinfoapi.cr [hunk 1]
```diff
@@ -1,9 +1,6 @@
 require "c/winbase"
 
 lib LibC
-  fun GetSystemTimeAsFileTime(time : FILETIME*)
-  fun GetSystemTimePreciseAsFileTime(time : FILETIME*)
```
Comment (straight-shoota) [inline src/lib_c/x86_64-windows-msvc/c/sysinfoapi.cr:5]: These functions are both defined in `sysinfoapi.h`, so it looks like they should stay here and the duplicates in `winbase.cr` be removed.

**Review by straight-shoota — APPROVED** [review]
Just one minor detail. Apart from that I'm good 🚀

File: src/lib_c/x86_64-windows-msvc/c/memoryapi.cr [hunk 2]
```diff
@@ -10,4 +10,17 @@ lib LibC
   MEM_RELEASE  = 0x8000
 
   fun VirtualFree(lpAddress : Void*, dwSize : SizeT, dwFreeType : DWORD) : BOOL
+
+  struct MEMORY_BASIC_INFORMATION
```
Comment (HertzDevil) [inline src/lib_c/x86_64-windows-msvc/c/memoryapi.cr:14]: This struct actually belongs in `winnt.h`

File: src/lib_c/x86_64-windows-msvc/c/processthreadsapi.cr [hunk 3]
```diff
@@ -32,9 +33,22 @@ lib LibC
     hStdError : HANDLE
   end
 
+  struct NT_TIB
```
Comment (HertzDevil) [inline src/lib_c/x86_64-windows-msvc/c/processthreadsapi.cr:36]: This struct actually belongs in `winnt.h`

**Review by HertzDevil — APPROVED** [review]

**Pull request merged**